src/libs/elektra/kdb.c

```diff
@@ -1324,6 +1381,441 @@ int kdbSet (KDB * handle, KeySet * ks, Key * parentKey)
+			      enum PluginContractState pluginState, KeySet * pluginConfig ELEKTRA_UNUSED, Key * errorKey ELEKTRA_UNUSED)
+{
+	switch (pluginState)
+	{
+	case PLUGIN_STATE_UNMOUNTED:
+		ksDel (pluginConfig);
+		return ensurePluginUnmounted (handle, mountpoint, pluginName, errorKey);
+	case PLUGIN_STATE_MOUNTED:
+		return 0; // TODO: ensurePluginMounted (handle, mountpoint, pluginName, pluginConfig, errorKey);
+	case PLUGIN_STATE_REMOUNT:
+		return 0; // TODO: ensurePluginUnmounted (handle, mountpoint, pluginName, errorKey) && ensurePluginMounted (handle,
+			  // mountpoint, pluginName, pluginConfig, errorKey);
+	default:
+		ELEKTRA_ASSERT (0, "missing switch case");
+		return 0;
+	}
+}
+
+/**
+ * This function can be used the given KDB @p handle meets certain conditions,
+ * specified in @p contract. Currently the following conditions are supported:
+ *
+ * - `system/plugins/<mountpoint>/<pluginname>` defines the state of the plugin
+ *   `<pluginname>` for the mountpoint `<mountpoint>`:
+ * 	- The value `unmounted` ensures the plugin is not mounted, at this mountpoint.
+ * 	- The value `mounted` ensures the plugin is mounted, at this mountpoint.
+ * 	  If the plugin is not mounted, we will try to mount it.
+ * 	- The value `remount` always mounts the plugin, at this mountpoint.
+ * 	  If it was already mounted, it will me unmounted and mounted again.
+ * 	  This can be used to ensure the plugin is mounted with a certain configuration.
+ * - Keys below `system/plugins/<mountpoint>/<pluginname>/config` are extracted and used
+ *   as the plugins config KeySet during mounting. `system/plugins/<mountpoint>/<pluginname>`
+ *   will be repleced by `user` in the keynames. If no keys are given, an empty KeySet is used.
+ *
+ * There are a few special values for `<mountpoint>`:
+ * - `global` is used to indicate the plugin should (un)mounted as a global plugin.
+ *   Currently this only supports (un)mounting plugins from/to the subposition `maxonce`.
+ * - `parent` is used to indicate the keyname of @p parentKey shall be used as the mountpoint.
+ *
+ * If `<mountpoint>` is none of those values, it has to be valid keyname with the slashes escaped.
+ * That means it has to start with `/`, `user`, `system`, `dir` or `spec`.
+ *
+ * If `<mountpoint>` is NOT `global`, currently only `unmounted` is supported (not `mounted` and `remounted`).
+ *
+ * NOTE: This function only works properly, if the list plugin is mounted in all global positions.
+ * If this is not the case, 1 will be returned, because this is seen as an implicit condition in the contract.
+ * Additionally any contract that specifies conditions for the list plugin is rejected as malformed.
+ *
+ * @param handle    contains internal information of @link kdbOpen() opened @endlink key database
+ * @param contract  KeySet containing the contract described above.
+ *                  Because of the implementation, this will always be `ksDel()`ed. <b>Even in error cases.</b>
```

Comment:
" Because of the implementation" is redundant information. Maybe you wanted to say: "To allow `kdbEnsure(.., ksNew(...)`..."

Reply:
> Maybe you wanted to say: "To allow kdbEnsure(.., ksNew(...)..."

That is a nice side effect, but in most cases won't make for good, readable code. The actual reason is that I use `ksCut` and wanted to avoid having to `ksDup` first, because the contract shouldn't be needed after calling `kdbEnsure` anyway.